pkg/storage/tsdb/bucketindex/writer.go

```diff
@@ -0,0 +1,273 @@
+		return nil, ErrBlockMetaNotFound
+	}
+	if err != nil {
+		return nil, errors.Wrapf(err, "get block meta file: %v", metaFile)
+	}
+	defer runutil.CloseWithLogOnErr(w.logger, r, "close get block meta file")
+
+	metaContent, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, errors.Wrapf(err, "read block meta file: %v", metaFile)
+	}
+
+	// Unmarshal it.
+	m := metadata.Meta{}
+	if err := json.Unmarshal(metaContent, &m); err != nil {
+		return nil, errors.Wrapf(ErrBlockMetaCorrupted, "unmarshal block meta file %s: %v", metaFile, err)
+	}
+
+	if m.Version != metadata.TSDBVersion1 {
+		return nil, errors.Errorf("unexpected block meta version: %s version: %d", metaFile, m.Version)
+	}
+
+	block := BlockFromThanosMeta(m)
+
+	// Get the meta.json attributes.
+	attrs, err := w.bkt.Attributes(ctx, metaFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "read meta file attributes: %v", metaFile)
+	}
+
+	// Since the meta.json file is the last file of a block being uploaded and it's immutable
+	// we can safely assume that the last modified timestamp of the meta.json is the time when
+	// the block has completed to be uploaded.
+	block.UploadedAt = attrs.LastModified.Unix()
+
+	return block, nil
+}
+
+func (w *Writer) generateBlockDeletionMarksIndex(ctx context.Context, old []*BlockDeletionMark) ([]*BlockDeletionMark, error) {
+	out := make([]*BlockDeletionMark, 0, len(old))
+	discovered := map[ulid.ULID]struct{}{}
+
+	// Find all markers in the storage.
+	err := w.bkt.Iter(ctx, MarkersPathname+"/", func(name string) error {
+		if blockID, ok := IsBlockDeletionMarkFilename(path.Base(name)); ok {
+			discovered[blockID] = struct{}{}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "list block deletion marks")
+	}
+
+	// Since deletion marks are immutable, all markers already existing in the index can just be copied.
+	for _, m := range old {
+		if _, ok := discovered[m.ID]; ok {
+			out = append(out, m)
+			delete(discovered, m.ID)
+		}
+	}
+
+	// Remaining markers are new ones and we have to fetch them.
+	for id := range discovered {
+		m, err := w.generateBlockDeletionMarkIndexEntry(ctx, id)
+		if errors.Is(err, ErrBlockDeletionMarkNotFound) {
+			// This could happen if the block is permanently deleted between the "list objects" and now.
+			level.Warn(w.logger).Log("msg", "skipped missing block deletion mark when generating bucket index", "block", id.String())
+			continue
+		}
+		if errors.Is(err, ErrBlockDeletionMarkCorrupted) {
+			level.Error(w.logger).Log("msg", "skipped corrupted block deletion mark when generating bucket index", "block", id.String(), "err", err)
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		out = append(out, m)
+	}
+
+	return out, nil
+}
+
+func (w *Writer) generateBlockDeletionMarkIndexEntry(ctx context.Context, id ulid.ULID) (*BlockDeletionMark, error) {
+	markFile := path.Join(id.String(), metadata.DeletionMarkFilename)
+
+	// Get the block's deletion mark file.
```

Comment:
We can replace marker reading code in this method with:

```
	m := metadata.DeletionMark{}
	err := metadata.ReadMarker(ctx, w.logger, w.bkt, id.String(), &m)
```

(`UserBucketClient` already implements required `InstrumentedBucket`)

Reply:
Very good point. Done, thanks!